File: src/vscripts/abilities/heroes/tmp/pango/pango_special_attack/pango_special_attack.lua
pango_special_attack = class({})
LinkLuaModifier("modifier_pango_special_attack_displacement", "abilities/heroes/pango/pango_special_attack/modifier_pango_special_attack_displacement", LUA_MODIFIER_MOTION_BOTH)


GetAnimation(){		return GameActivity.DOTA_FORCESTAFF_END }
GetPlaybackRateOverride(){ 	    return 0.5 }

GetCastingCrawl(){
    return 20
}

function pango_special_attack:GetCastRange(vLocation, target)
    if(this.caster.HasModifier("modifier_pango_mobility") ) {
        return this.BaseClass.GetCastRange(this, vLocation, target) * 1.2
    }
    return this.BaseClass.GetCastRange(this, vLocation, target)
}

GetCastPoint(){
    if(this.caster.HasModifier("modifier_pango_mobility") ) {
        return 0.0
    }
    return super.GetCastPoint()
}


function pango_special_attack:HasPriority()
    return true
}

OnSpellStart(){
	
	const origin = this.caster.GetAbsOrigin()
	const point = CustomAbilitiesLegacy:GetCursorPosition(this)

	const direction = (point - origin):Normalized()
    const distance = this.GetCastRange(Vector(0,0,0), undefined)
    const air_time = 0.4

    const modifier_pango_mobility = CustomEntitiesLegacy:SafeGetModifier(this.caster, "modifier_pango_mobility")
    if(modifier_pango_mobility ) {
        modifier_pango_mobility:SetDuration(modifier_pango_mobility:GetRemainingTime() + air_time, true)
        direction = CustomEntitiesLegacy:GetDirection(this.caster)
    }

    this.caster.AddNewModifier(
        this.caster, -- player source
        this, -- ability source
        "modifier_pango_special_attack_displacement", -- modifier name
        {
            x = direction.x,
            y = direction.y,
            r = distance,
            speed = distance/air_time,
            peak = 250,
        }
    )

    EmitSoundOn("Hero_Pangolier.TailThump.Cast", this.caster)
}

function pango_special_attack:Crash(iRadius)
    
    const origin = this.caster.GetAbsOrigin()
    const damage_block = this.GetSpecialValueFor("damage_block")
    const shield_duration = this.GetSpecialValueFor("duration")
    const shield_providers = 0
    const damage_table = {
        attacker = this.caster,
        damage = this.GetSpecialValueFor("ability_damage"),
        damage_type = DamageTypes.PURE,
    }
	ApplyCallbackForUnitsInArea(this.caster, origin, iRadius, DOTA_UNIT_TARGET_TEAM_ENEMY, function(unit)
		damage_table.victim = unit
		ApplyDamage(damage_table)

		if(! CustomEntitiesLegacy:IsObstacle(unit) ) {
			shield_providers = shield_providers + 1
		}
	})

	if(shield_providers ~= 0 ) {
		this.caster.AddNewModifier(this.caster, this, "modifier_shield", { duration = shield_duration, damage_block = damage_block * shield_providers })
	}

	EFX("particles/units/heroes/hero_pangolier/pangolier_tailthump.vpcf", ParticleAttachment.WORLDORIGIN, undefined, {
		cp0 = origin
	})
	CreateRadiusMarker(this.caster, origin, iRadius, RADIUS_SCOPE_PUBLIC, 0.1)
	EmitSoundOn("Hero_Pangolier.TailThump", this.caster)
}

if(IsClient() ) { require("wrappers/abilities") }
Abilities.Castpoint(pango_special_attack)

File: src/vscripts/abilities/heroes/pango/pango_special_attack.ts
import { createRadiusMarker, findUnitsInRadius } from "../../../util";
import { registerAbility, registerModifier } from "../../../lib/dota_ts_adapter";
import { ModifierDisplacement } from "../../../modifiers/modifier_displacement";
import { ModifierShield } from "../../../modifiers/modifier_shield";
import { direction2D, getCursorPosition, isObstacle } from "../../../util";
import { CustomAbility } from "../../framework/custom_ability";
import { ModifierPangoMobility } from "./modifier_pango_mobility";
import { PangoSecondAttack } from "./pango_second_attack";

@registerAbility("pango_special_attack")
export class PangoSpecialAttack extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_FORCESTAFF_END;
    }

    GetPlaybackRateOverride() {
        return 0.5;
    }

    GetCastingCrawl() {
        return 20;
    }

    GetCastRange(location: Vector, target?: CDOTA_BaseNPC) {
        if (ModifierPangoMobility.findOne(this.caster) && IsServer()) {
            return super.GetCastRange(location, target) * 1.2;
        }
        return super.GetCastRange(location, target);
    }

    GetCastPoint() {
        if (ModifierPangoMobility.findOne(this.caster)) {
            return 0.0;
        }
        return super.GetCastPoint();
    }

    // function pango_special_attack:HasPriority()
    //     return true
    // }

    OnSpellStart() {
        const origin = this.caster.GetAbsOrigin();
        const point = getCursorPosition(this.caster);
        const distance = this.GetCastRange(Vector(0, 0, 0), undefined);
        const airTime = 0.4;

        const modifier = ModifierPangoMobility.findOne(this.caster);
        modifier?.SetDuration(modifier?.GetRemainingTime() + airTime, true);
        const direction = modifier ? CustomEntitiesLegacy.GetDirection(this.caster) : direction2D(origin, point);

        ModifierPangoSpecialAttack.apply(this.caster, this.caster, this, {
            x: direction.x,
            y: direction.y,
            distance: distance,
            speed: distance / airTime,
            peak: 250
        });

        EmitSoundOn("Hero_Pangolier.TailThump.Cast", this.caster);
    }

    Crash(radius: number) {
        const origin = this.caster.GetAbsOrigin();
        const damageBlock = this.GetSpecialValueFor("damage_block");
        const shieldDuration = this.GetSpecialValueFor("duration");
        const enemies = findUnitsInRadius(
            this.caster,
            origin,
            radius,
            UnitTargetTeam.ENEMY,
            UnitTargetType.HERO + UnitTargetType.BASIC,
            UnitTargetFlags.NONE,
            FindOrder.ANY
        );

        const shieldProviders = enemies.filter(enemy => !isObstacle(enemy)).length;

        for (const enemy of enemies) {
            ApplyDamage({
                victim: enemy,
                attacker: this.caster,
                damage: this.GetSpecialValueFor("ability_damage"),
                damage_type: DamageTypes.PURE
            });
        }

        if (shieldProviders > 0) {
            ModifierShield.apply(this.caster, this.caster, undefined, {
                duration: shieldDuration,
                damageBlock: damageBlock * shieldProviders
            });
        }

        EFX("particles/units/heroes/hero_pangolier/pangolier_tailthump.vpcf", ParticleAttachment.WORLDORIGIN, undefined, {
            cp0: origin
        });
        createRadiusMarker(this.caster, origin, radius, "public", 0.1);
        EmitSoundOn("Hero_Pangolier.TailThump", this.caster);
    }
}

@registerModifier()
class ModifierPangoSpecialAttack extends ModifierDisplacement<PangoSpecialAttack> {
    OnDestroy() {
        super.OnDestroy();

        if (IsServer()) {
            this.ability.Crash(this.radius);

            if (this.ability.GetLevel() >= 2) {
                const origin = this.caster.GetAbsOrigin();
                const direction = this.caster.GetForwardVector();
                const pangoSecondAttack = PangoSecondAttack.findOne(this.caster);
                const range = (pangoSecondAttack?.GetCastRange(Vector(0, 0, 0), undefined) ?? 1) / 2;
                pangoSecondAttack?.Cut(origin, direction, range);
                pangoSecondAttack?.Cut(origin, direction.__mul(-1), range);
            }
        }
    }

    DeclareFunctions() {
        return [ModifierFunction.OVERRIDE_ANIMATION, ModifierFunction.OVERRIDE_ANIMATION_RATE];
    }

    GetOverrideAnimation() {
        if (ModifierPangoMobility.findOne(this.parent)) {
            return GameActivity.IDLE;
        }
        return GameActivity.DOTA_CAST_ABILITY_2;
    }

    GetOverrideAnimationRate() {
        return 1.0;
    }

    CheckState() {
        return {
            ...super.CheckState(),
            [ModifierState.NO_HEALTH_BAR]: true,
            [ModifierState.INVULNERABLE]: true,
            [ModifierState.OUT_OF_GAME]: true,
            [ModifierState.NO_UNIT_COLLISION]: true
        };
    }
}
